Initial commit

agnivade opened the pull request:
Logging runtime stats using pg_stat_statements

Review request: @isacikgoz @streamer45

agnivade committed:
Initial commit

Logging runtime stats using pg_stat_statements

streamer45 commented on main.go:
```diff
@@ -0,0 +1,114 @@
+			return
+		}
+
+		_, err = conn.Exec(context.Background(), fmt.Sprintf("%s", buf))
```
Would passing a simple `string(buf)` work or were you planning on adding formatting?

agnivade replied:
IIRC `string(buf)` behaves strangely with unprintable characters, but in this case, it should be fine I guess.

isacikgoz commented on main.go:
```diff
@@ -0,0 +1,114 @@
+			continue
+		}
+
+		split := strings.Split(entry.Name(), "_")
+		if len(split) < 2 {
+			logger.Printf("_ separator not found for this migration %s. Skipping ..\n", entry.Name())
+			continue
+		}
```
We can use https://pkg.go.dev/github.com/mattermost/morph/models#pkg-variables `Regex` for this. In fact you can use https://pkg.go.dev/github.com/mattermost/morph/models#NewMigration to parse the migration.

agnivade replied:
Ah very nice.

isacikgoz reviewed (approved):
LGTM, added some morph references to avoid some labor work.

streamer45 commented on main.go:
```diff
@@ -0,0 +1,114 @@
+		return
+	}
+
+	conn, err := pgx.Connect(context.Background(), dsn)
```
Are we okay with not using a timeout on the context here?

agnivade replied:
Yes, in a future PR :)

streamer45 commented on main.go:
```diff
@@ -0,0 +1,114 @@
+)
+
+func main() {
+	var startingMigration int
+	var sourceDir, dsn string
+	flag.IntVar(&startingMigration, "start", 0, "Starting migration number from where to apply migrations")
+	flag.StringVar(&sourceDir, "source_dir", "", "Source directory containing the migrations. Example: ~/mattermost-server/server/channels/db/migrations/postgres")
+	flag.StringVar(&dsn, "dsn", "postgres://mmuser:mostest@localhost/mattermost_test?sslmode=disable", "Database DSN to connect to")
+	flag.Parse()
```
Purely a note for later but since we are actually applying migrations it would be good to add some sort of disclaimer to make extra sure the users of this tool understand what they are about to do.

agnivade replied:
Makes sense.

streamer45 commented on main.go:
```diff
@@ -0,0 +1,114 @@
+		var numRows int
+		pgx.ForEachRow(rows, []any{&query, &execTime, &numRows}, func() error {
+			// Ignore the reset query
+			if query == "SELECT pg_stat_statements_reset()" {
```
I wonder if it would make sense to extract these queries into constants. Also I guess the terminating semicolon doesn't matter?

agnivade replied:
> I wonder if it would make sense to extract these queries into constants.

I think it's fine for now.

> Also I guess the terminating semicolon doesn't matter?

No, the output from pg_stat_statements remove the semicolon from the query.

streamer45 reviewed (approved):
Looks good, purely left some minor thoughts.

agnivade committed:
MM-51701: Show lock output (#2)

Example run
```
Starting migration:                        000106_fileinfo_channelid.up.sql---------------------------------------------
LOCKS:
LOCK TYPE: relation, RELATION: fileinfo, MODE: RowExclusiveLock, GRANTED: true
LOCK TYPE: relation, RELATION: posts, MODE: AccessShareLock, GRANTED: true
LOCK TYPE: relation, RELATION: fileinfo, MODE: ShareLock, GRANTED: true
LOCK TYPE: relation, RELATION: fileinfo, MODE: AccessExclusiveLock, GRANTED: true
QUERY:  UPDATE fileinfo SET channelid = posts.channelid FROM posts WHERE fileinfo.channelid IS NULL AND fileinfo.postid = posts.id
EXEC TIME: 0.012882 ms, ROWS AFFECTED: 0
QUERY:  CREATE INDEX IF NOT EXISTS idx_fileinfo_channel_id_create_at ON fileinfo(channelid, createat)
EXEC TIME: 2088.678447 ms, ROWS AFFECTED: 0
QUERY:  SELECT locktype, relation::regclass, mode, granted
                                FROM pg_catalog.pg_locks l
                                LEFT JOIN pg_catalog.pg_database db ON db.oid = l.database
                                JOIN pg_class c ON l.relation=c.oid
                                WHERE (db.datname = current_database() OR db.datname IS NULL) AND c.relkind=$1 AND NOT pid = pg_backend_pid() AND locktype=$2
EXEC TIME: 0.793934 ms, ROWS AFFECTED: 9
QUERY:  ALTER TABLE fileinfo ADD COLUMN IF NOT EXISTS channelid varchar(26)
EXEC TIME: 0.829388 ms, ROWS AFFECTED: 0
```

https://mattermost.atlassian.net/browse/MM-51701

agnivade committed:
Review comments
